From a71867dec990e0fb5f429574df14e3ce83c25e84 Mon Sep 17 00:00:00 2001
From: jhenrique09 <jhsv09@gmail.com>
Date: Thu, 10 Feb 2022 22:27:30 -0300
Subject: [PATCH] Import of Face Unlock for T (2/2)

Squash of:
- Initial import of Face Unlock for T
- Settings: Allow using face as auth method for apps
- [2/2] Allow changing face unlock method when locked
- Settings: Disable face enroll on Setup Wizard

Change-Id: I178b63d73f14742cf890cb05386ad4e82fd8c2be
---
 Android.bp                                    |   1 +
 AndroidManifest.xml                           |  15 ++
 res/layout/face_enroll_button.xml             |  33 ----
 .../face_enroll_introduction_invisible.xml    | 164 ++++++++++++++++++
 res/layout/face_remove_button.xml             |  33 ----
 res/values/arrays.xml                         |   9 +
 res/values/strings.xml                        |   8 +
 res/xml/configure_notification_settings.xml   |  14 +-
 res/xml/security_lockscreen_settings.xml      |  11 ++
 .../security_settings_combined_biometric.xml  |   4 +-
 res/xml/security_settings_face.xml            |  42 +++--
 .../biometrics/BiometricEnrollActivity.java   |   5 +
 .../BiometricEnrollIntroduction.java          |   2 +-
 ...metricSettingsAppPreferenceController.java |   4 +-
 ...cSettingsKeyguardPreferenceController.java |   4 +-
 ...iometricWaysToUsePreferenceController.java |  35 ++++
 .../face/FaceEnrollIntroduction.java          | 128 +++++++++++---
 ...FaceProfileStatusPreferenceController.java |   5 +
 .../biometrics/face/FaceSettings.java         |  27 ++-
 .../FaceSettingsAppPreferenceController.java  |   3 +-
 ...SettingsAttentionPreferenceController.java |   4 +-
 ...ceSettingsConfirmPreferenceController.java |   5 +
 ...tingsEnrollButtonPreferenceController.java |  19 +-
 ...eSettingsKeyguardPreferenceController.java |   4 +-
 ...sLockscreenBypassPreferenceController.java |  11 +-
 ...creenUnlockMethodPreferenceController.java |  59 +++++++
 ...tingsRemoveButtonPreferenceController.java |  50 ++----
 .../face/FaceStatusPreferenceController.java  |   6 +
 .../biometrics/face/FaceStatusUtils.java      |   5 +
 ...aceUnlockCategoryPreferenceController.java |   4 +-
 .../settings/custom/ListPreference.java       |  55 ++++++
 .../custom/SecureSettingListPreference.java   |  49 ++++++
 .../settings/custom/SecureSettingsStore.java  |  74 ++++++++
 .../settings/custom/biometrics/FaceUtils.java |  31 ++++
 .../biometrics/face/FaceEnrollActivity.java   |  34 ++++
 .../FaceSettingsRedoPreferenceController.java | 138 +++++++++++++++
 .../contextualcards/FaceReEnrollDialog.java   |   7 +-
 37 files changed, 931 insertions(+), 171 deletions(-)
 delete mode 100644 res/layout/face_enroll_button.xml
 create mode 100644 res/layout/face_enroll_introduction_invisible.xml
 delete mode 100644 res/layout/face_remove_button.xml
 create mode 100644 src/com/android/settings/biometrics/combination/BiometricWaysToUsePreferenceController.java
 create mode 100644 src/com/android/settings/biometrics/face/FaceSettingsLockscreenUnlockMethodPreferenceController.java
 create mode 100644 src/com/android/settings/custom/ListPreference.java
 create mode 100644 src/com/android/settings/custom/SecureSettingListPreference.java
 create mode 100644 src/com/android/settings/custom/SecureSettingsStore.java
 create mode 100644 src/com/android/settings/custom/biometrics/FaceUtils.java
 create mode 100644 src/com/android/settings/custom/biometrics/face/FaceEnrollActivity.java
 create mode 100644 src/com/android/settings/custom/biometrics/face/FaceSettingsRedoPreferenceController.java

diff --git a/Android.bp b/Android.bp
index f56ca95f9b..f601380b89 100644
--- a/Android.bp
+++ b/Android.bp
@@ -89,6 +89,7 @@ android_library {
         "SettingsLibActivityEmbedding",
         "Settings-change-ids",
         "SystemUIUnfoldLib",
+        "faceunlock_framework",
     ],
 
     libs: [
diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index a4d899d5dc..eb836a7ba8 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -120,6 +120,10 @@
     <uses-permission android:name="android.permission.START_VIEW_APP_FEATURES" />
     <uses-permission android:name="android.permission.CUSTOMIZE_SYSTEM_UI" />
 
+    <!-- Face Unlock -->
+    <uses-permission android:name="com.android.settings.FACE_UNLOCK"/>
+    <permission android:name="com.android.settings.FACE_UNLOCK" android:protectionLevel="signatureOrSystem"/>
+
     <application
             android:name=".SettingsApplication"
             android:label="@string/settings_label"
@@ -4545,6 +4549,17 @@
             </intent-filter>
         </activity>
 
+        <!-- Face unlock -->
+        <activity android:name="com.android.settings.custom.biometrics.face.FaceEnrollActivity"
+            android:permission="com.android.settings.FACE_UNLOCK"
+            android:theme="@style/Transparent"
+            android:exported="true">
+            <intent-filter android:priority="1">
+                <action android:name="com.android.settings.intent.action.FACE_ENROLL"/>
+                <category android:name="android.intent.category.DEFAULT"/>
+            </intent-filter>
+        </activity>
+
         <!-- This is the longest AndroidManifest.xml ever. -->
     </application>
 </manifest>
diff --git a/res/layout/face_enroll_button.xml b/res/layout/face_enroll_button.xml
deleted file mode 100644
index 6266650349..0000000000
--- a/res/layout/face_enroll_button.xml
+++ /dev/null
@@ -1,33 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
-  ~ Copyright (C) 2019 The Android Open Source Project
-  ~
-  ~ Licensed under the Apache License, Version 2.0 (the "License");
-  ~ you may not use this file except in compliance with the License.
-  ~ You may obtain a copy of the License at
-  ~
-  ~      http://www.apache.org/licenses/LICENSE-2.0
-  ~
-  ~ Unless required by applicable law or agreed to in writing, software
-  ~ distributed under the License is distributed on an "AS IS" BASIS,
-  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-  ~ See the License for the specific language governing permissions and
-  ~ limitations under the License
-  -->
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    android:layout_width="match_parent"
-    android:layout_height="wrap_content"
-    android:minHeight="?android:attr/listPreferredItemHeight"
-    android:paddingEnd="?android:attr/listPreferredItemPaddingEnd"
-    android:paddingStart="?android:attr/listPreferredItemPaddingStart">
-
-    <Button
-        android:id="@+id/security_settings_face_settings_enroll_button"
-        style="@style/SudGlifButton.Primary"
-        android:layout_width="wrap_content"
-        android:layout_height="wrap_content"
-        android:layout_gravity="start"
-        android:text="@string/security_settings_face_settings_enroll"/>
-
-</LinearLayout>
\ No newline at end of file
diff --git a/res/layout/face_enroll_introduction_invisible.xml b/res/layout/face_enroll_introduction_invisible.xml
new file mode 100644
index 0000000000..217cacd97c
--- /dev/null
+++ b/res/layout/face_enroll_introduction_invisible.xml
@@ -0,0 +1,164 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (C) 2021 The Android Open Source Project
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~      http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<com.google.android.setupdesign.GlifLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    style="?attr/face_layout_theme"
+    android:id="@+id/setup_wizard_layout"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    app:sudDescriptionText="@string/security_settings_face_enroll_introduction_message"
+    android:visibility="invisible">
+
+    <LinearLayout
+        style="@style/SudContentFrame"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:clipChildren="false"
+        android:clipToPadding="false"
+        android:orientation="vertical"
+        android:layout_marginStart="?attr/sudMarginStart"
+        android:layout_marginEnd="?attr/sudMarginEnd">
+
+        <com.google.android.setupdesign.view.RichTextView
+            android:id="@+id/error_text"
+            style="@style/SudDescription.Glif"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:visibility="gone"/>
+
+        <FrameLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
+
+            <ImageView
+                style="@style/SudContentIllustration"
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:contentDescription="@null"
+                android:src="@drawable/face_enroll_intro_illustration"/>
+
+        </FrameLayout>
+
+        <!-- Contains the extra information text at the bottom -->
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+
+            <!-- Keep in mind -->
+            <TextView
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                style="@style/BiometricEnrollIntroTitle"
+                android:text="@string/security_settings_face_enroll_introduction_info_title" />
+
+            <LinearLayout
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:orientation="horizontal">
+
+                <ImageView
+                    android:id="@+id/icon_glasses"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:background="@drawable/ic_face_enroll_introduction_glasses"/>
+                <Space
+                    android:layout_width="16dp"
+                    android:layout_height="wrap_content"/>
+                <TextView
+                    android:id="@+id/info_message_glasses"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    style="@style/BiometricEnrollIntroMessage" />
+            </LinearLayout>
+
+            <LinearLayout
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:orientation="horizontal">
+
+                <ImageView
+                    android:id="@+id/icon_looking"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:background="@drawable/ic_face_enroll_introduction_visibility"/>
+                <Space
+                    android:layout_width="16dp"
+                    android:layout_height="wrap_content"/>
+                <TextView
+                    android:id="@+id/info_message_looking"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    style="@style/BiometricEnrollIntroMessage" />
+            </LinearLayout>
+
+            <LinearLayout
+                android:id="@+id/info_row_require_eyes"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:orientation="horizontal"
+                android:visibility="gone">
+
+                <ImageView
+                    android:id="@+id/icon_require_eyes"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:background="@drawable/ic_settings_24dp"/>
+                <Space
+                    android:layout_width="16dp"
+                    android:layout_height="wrap_content"/>
+                <TextView
+                    android:id="@+id/info_message_require_eyes"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    style="@style/BiometricEnrollIntroMessage" />
+            </LinearLayout>
+
+            <!-- How it works -->
+            <TextView
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                style="@style/BiometricEnrollIntroTitle"
+                android:text="@string/security_settings_face_enroll_introduction_how_title" />
+
+            <TextView
+                android:id="@+id/how_message"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                style="@style/BiometricEnrollIntroMessage" />
+
+            <!-- You're in control -->
+            <TextView
+                android:id="@+id/title_in_control"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                style="@style/BiometricEnrollIntroTitle" />
+
+            <TextView
+                android:id="@+id/message_in_control"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                style="@style/BiometricEnrollIntroMessage"
+                android:paddingBottom="0dp" />
+
+        </LinearLayout>
+
+    </LinearLayout>
+
+</com.google.android.setupdesign.GlifLayout>
\ No newline at end of file
diff --git a/res/layout/face_remove_button.xml b/res/layout/face_remove_button.xml
deleted file mode 100644
index 2c2497ab1c..0000000000
--- a/res/layout/face_remove_button.xml
+++ /dev/null
@@ -1,33 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
-  ~ Copyright (C) 2018 The Android Open Source Project
-  ~
-  ~ Licensed under the Apache License, Version 2.0 (the "License");
-  ~ you may not use this file except in compliance with the License.
-  ~ You may obtain a copy of the License at
-  ~
-  ~      http://www.apache.org/licenses/LICENSE-2.0
-  ~
-  ~ Unless required by applicable law or agreed to in writing, software
-  ~ distributed under the License is distributed on an "AS IS" BASIS,
-  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-  ~ See the License for the specific language governing permissions and
-  ~ limitations under the License.
-  -->
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    android:layout_width="match_parent"
-    android:layout_height="wrap_content"
-    android:minHeight="?android:attr/listPreferredItemHeight"
-    android:paddingEnd="?android:attr/listPreferredItemPaddingEnd"
-    android:paddingStart="?android:attr/listPreferredItemPaddingStart">
-
-    <Button
-        android:id="@+id/security_settings_face_settings_remove_button"
-        style="@style/SudGlifButton.Primary"
-        android:layout_width="wrap_content"
-        android:layout_height="wrap_content"
-        android:layout_gravity="start"
-        android:text="@string/security_settings_face_settings_remove_face_model"/>
-
-</LinearLayout>
\ No newline at end of file
diff --git a/res/values/arrays.xml b/res/values/arrays.xml
index 9bc29560c9..7557324f02 100644
--- a/res/values/arrays.xml
+++ b/res/values/arrays.xml
@@ -1659,4 +1659,13 @@
     <integer-array name="network_mode_3g_deprecated_carrier_id" translatable="false">
     </integer-array>
 
+    <!-- Face Unlock -->
+    <string-array name="face_unlock_method_entries">
+        <item>@string/face_unlock_method_default</item>
+        <item>@string/face_unlock_method_swipe</item>
+    </string-array>
+    <string-array name="face_unlock_method_values">
+        <item>0</item>
+        <item>1</item>
+    </string-array>
 </resources>
diff --git a/res/values/strings.xml b/res/values/strings.xml
index 6ca21ee95b..8b99dde6ea 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -14551,4 +14551,12 @@
     <!-- [CHAR LIMIT=NONE] Hint for QR code process failure -->
     <string name="bt_le_audio_qr_code_is_not_valid_format">QR code isn\u0027t a valid format</string>
 
+    <!-- Face Unlock -->
+    <string name="security_settings_face_unlock_redo_face_scan_summary">Remove existing face scan and create a new scan</string>
+    <string name="security_settings_face_unlock_redo_face_scan_title">Redo face scan</string>
+    <string name="face_redo_confirm_btn">Redo</string>
+    <string name="face_redo_warning_msg">Do you want to remove the existing face scan and create a new one?</string>
+    <string name="face_unlock_method_title">Allow face unlock on lock screen</string>
+    <string name="face_unlock_method_default">When the screen is turned on</string>
+    <string name="face_unlock_method_swipe">When swiping on lock screen</string>
 </resources>
diff --git a/res/xml/configure_notification_settings.xml b/res/xml/configure_notification_settings.xml
index 27d5760992..c6767cd55c 100644
--- a/res/xml/configure_notification_settings.xml
+++ b/res/xml/configure_notification_settings.xml
@@ -98,9 +98,21 @@
             android:summary="@string/lock_screen_notifs_redact_work_summary"
             settings:controller="com.android.settings.notification.RedactNotificationPreferenceController" />
 
+        <com.android.settings.custom.SecureSettingListPreference
+            android:key="face_unlock_method"
+            android:order="16"
+            android:title="@string/face_unlock_method_title"
+            android:summary="%s"
+            android:dialogTitle="@string/face_unlock_method_title"
+            android:entries="@array/face_unlock_method_entries"
+            android:entryValues="@array/face_unlock_method_values"
+            android:defaultValue="0"
+            settings:searchable="false"
+            settings:controller="com.android.settings.biometrics.face.FaceSettingsLockscreenUnlockMethodPreferenceController" />
+
         <SwitchPreference
             android:key="notification_lockscreen_bypass"
-            android:order="16"
+            android:order="17"
             android:title="@string/lockscreen_bypass_title"
             android:summary="@string/lockscreen_bypass_summary"
             settings:searchable="false"
diff --git a/res/xml/security_lockscreen_settings.xml b/res/xml/security_lockscreen_settings.xml
index 80e8fe69d2..f45d462fed 100644
--- a/res/xml/security_lockscreen_settings.xml
+++ b/res/xml/security_lockscreen_settings.xml
@@ -28,6 +28,17 @@
             android:summary="@string/summary_placeholder"
             settings:keywords="@string/keywords_lock_screen_notif"/>
 
+        <com.android.settings.custom.SecureSettingListPreference
+            android:key="face_unlock_method"
+            android:title="@string/face_unlock_method_title"
+            android:summary="%s"
+            android:dialogTitle="@string/face_unlock_method_title"
+            android:entries="@array/face_unlock_method_entries"
+            android:entryValues="@array/face_unlock_method_values"
+            android:defaultValue="0"
+            settings:searchable="false"
+            settings:controller="com.android.settings.biometrics.face.FaceSettingsLockscreenUnlockMethodPreferenceController" />
+
         <SwitchPreference
             android:key="security_display_lockscreen_bypass"
             android:title="@string/lockscreen_bypass_title"
diff --git a/res/xml/security_settings_combined_biometric.xml b/res/xml/security_settings_combined_biometric.xml
index ef3a3fd1ff..f7d732bbb4 100644
--- a/res/xml/security_settings_combined_biometric.xml
+++ b/res/xml/security_settings_combined_biometric.xml
@@ -43,8 +43,8 @@
     </PreferenceCategory>
 
     <PreferenceCategory
-        android:key="biometric_ways_to_use">
-
+        android:key="biometric_ways_to_use"
+        settings:controller="com.android.settings.biometrics.combination.BiometricWaysToUsePreferenceController">
         <com.android.settingslib.RestrictedSwitchPreference
             android:key="biometric_settings_biometric_keyguard"
             android:title="@string/biometric_settings_use_biometric_unlock_phone"
diff --git a/res/xml/security_settings_face.xml b/res/xml/security_settings_face.xml
index f0d03507aa..d1f972eb1c 100644
--- a/res/xml/security_settings_face.xml
+++ b/res/xml/security_settings_face.xml
@@ -33,12 +33,6 @@
             android:title="@string/biometric_settings_use_biometric_for_apps"
             settings:keywords="@string/keywords_face_unlock"
             settings:controller="com.android.settings.biometrics.face.FaceSettingsAppPreferenceController"/>
-        <com.android.settingslib.RestrictedSwitchPreference
-            android:key="security_lockscreen_bypass"
-            android:title="@string/lockscreen_bypass_title"
-            android:summary="@string/lockscreen_bypass_summary"
-            settings:keywords="@string/keywords_lockscreen_bypass"
-            settings:controller="com.android.settings.biometrics.face.FaceSettingsLockscreenBypassPreferenceController" />
     </PreferenceCategory>
 
     <PreferenceCategory
@@ -58,25 +52,41 @@
             settings:keywords="@string/keywords_face_unlock"
             settings:controller="com.android.settings.biometrics.face.FaceSettingsConfirmPreferenceController"/>
 
+        <com.android.settings.custom.SecureSettingListPreference
+            android:key="face_unlock_method"
+            android:title="@string/face_unlock_method_title"
+            android:summary="%s"
+            android:dialogTitle="@string/face_unlock_method_title"
+            android:entries="@array/face_unlock_method_entries"
+            android:entryValues="@array/face_unlock_method_values"
+            android:defaultValue="0"
+            settings:searchable="false"
+            settings:controller="com.android.settings.biometrics.face.FaceSettingsLockscreenUnlockMethodPreferenceController" />
+
         <com.android.settingslib.RestrictedSwitchPreference
-            android:key="biometric_settings_lockscreen_bypass"
+            android:key="security_lockscreen_bypass"
             android:title="@string/lockscreen_bypass_title"
             android:summary="@string/lockscreen_bypass_summary"
             settings:keywords="@string/keywords_lockscreen_bypass"
-            settings:controller="com.android.settings.biometrics.face.BiometricLockscreenBypassPreferenceController" />
+            settings:controller="com.android.settings.biometrics.face.FaceSettingsLockscreenBypassPreferenceController" />
     </PreferenceCategory>
 
-    <com.android.settingslib.widget.LayoutPreference
-        android:key="security_settings_face_delete_faces_container"
+    <Preference
+        android:title="@string/security_settings_face_unlock_redo_face_scan_title"
+        android:key="security_settings_face_redo_face_scan"
+        android:summary="@string/security_settings_face_unlock_redo_face_scan_summary"
+        settings:controller="com.android.settings.custom.biometrics.face.FaceSettingsRedoPreferenceController"
+        settings:keywords="@string/keywords_face_unlock" />
+
+    <Preference
         android:title="@string/security_settings_face_settings_remove_face_model"
-        android:selectable="false"
-        android:layout="@layout/face_remove_button"/>
+        android:key="security_settings_face_delete_faces_container"
+        settings:keywords="@string/keywords_face_unlock"/>
 
-    <com.android.settingslib.widget.LayoutPreference
-        android:key="security_settings_face_enroll_faces_container"
+    <Preference
         android:title="@string/security_settings_face_settings_enroll"
-        android:selectable="false"
-        android:layout="@layout/face_enroll_button"/>
+        android:key="security_settings_face_enroll_faces_container"
+        settings:keywords="@string/keywords_face_unlock"/>
 
     <com.android.settingslib.widget.FooterPreference
         android:key="security_face_footer"
diff --git a/src/com/android/settings/biometrics/BiometricEnrollActivity.java b/src/com/android/settings/biometrics/BiometricEnrollActivity.java
index bb16f0bb9b..6f1007a911 100644
--- a/src/com/android/settings/biometrics/BiometricEnrollActivity.java
+++ b/src/com/android/settings/biometrics/BiometricEnrollActivity.java
@@ -57,6 +57,8 @@ import com.google.android.setupcompat.util.WizardManagerHelper;
 
 import java.util.List;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Trampoline activity launched by the {@code android.settings.BIOMETRIC_ENROLL} action which
  * shows the user an appropriate enrollment flow depending on the device's biometric hardware.
@@ -228,6 +230,9 @@ public class BiometricEnrollActivity extends InstrumentedActivity {
                 }
             }
         }
+        if (FaceUtils.isFaceUnlockSupported() && isSetupWizard){
+            mIsFaceEnrollable = false;
+        }
         if (mHasFeatureFingerprint) {
             final FingerprintManager fpManager = getSystemService(FingerprintManager.class);
             final List<FingerprintSensorPropertiesInternal> fpProperties =
diff --git a/src/com/android/settings/biometrics/BiometricEnrollIntroduction.java b/src/com/android/settings/biometrics/BiometricEnrollIntroduction.java
index cdd2d8b88d..204e1eb1e1 100644
--- a/src/com/android/settings/biometrics/BiometricEnrollIntroduction.java
+++ b/src/com/android/settings/biometrics/BiometricEnrollIntroduction.java
@@ -59,7 +59,7 @@ public abstract class BiometricEnrollIntroduction extends BiometricEnrollBase
     private static final String KEY_SCROLLED_TO_BOTTOM = "scrolled";
 
     private UserManager mUserManager;
-    private boolean mHasPassword;
+    protected boolean mHasPassword;
     private boolean mBiometricUnlockDisabledByAdmin;
     private TextView mErrorText;
     protected boolean mConfirmingCredentials;
diff --git a/src/com/android/settings/biometrics/combination/BiometricSettingsAppPreferenceController.java b/src/com/android/settings/biometrics/combination/BiometricSettingsAppPreferenceController.java
index a46ae7a728..29618f2018 100644
--- a/src/com/android/settings/biometrics/combination/BiometricSettingsAppPreferenceController.java
+++ b/src/com/android/settings/biometrics/combination/BiometricSettingsAppPreferenceController.java
@@ -28,6 +28,8 @@ import com.android.settings.core.TogglePreferenceController;
 import com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
 import com.android.settingslib.RestrictedLockUtilsInternal;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller that controls whether the biometrics authentication to be used in apps.
  */
@@ -69,7 +71,7 @@ public class BiometricSettingsAppPreferenceController extends TogglePreferenceCo
 
     @Override
     public int getAvailabilityStatus() {
-        if (!Utils.isMultipleBiometricsSupported(mContext)) {
+        if (!Utils.isMultipleBiometricsSupported(mContext) || FaceUtils.isFaceUnlockSupported()) {
             return UNSUPPORTED_ON_DEVICE;
         }
         if (mFaceManager == null || mFingerprintManager == null) {
diff --git a/src/com/android/settings/biometrics/combination/BiometricSettingsKeyguardPreferenceController.java b/src/com/android/settings/biometrics/combination/BiometricSettingsKeyguardPreferenceController.java
index 2d2255805c..7ad1c77479 100644
--- a/src/com/android/settings/biometrics/combination/BiometricSettingsKeyguardPreferenceController.java
+++ b/src/com/android/settings/biometrics/combination/BiometricSettingsKeyguardPreferenceController.java
@@ -26,6 +26,8 @@ import com.android.settings.core.TogglePreferenceController;
 import com.android.settingslib.RestrictedLockUtils;
 import com.android.settingslib.RestrictedLockUtilsInternal;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller that controls whether the biometrics authentication to unlock the device.
  */
@@ -63,7 +65,7 @@ public class BiometricSettingsKeyguardPreferenceController extends TogglePrefere
 
     @Override
     public int getAvailabilityStatus() {
-        if (!Utils.isMultipleBiometricsSupported(mContext)) {
+        if (FaceUtils.isFaceUnlockSupported() || !Utils.isMultipleBiometricsSupported(mContext)) {
             return UNSUPPORTED_ON_DEVICE;
         }
         return getRestrictingAdmin() != null ? DISABLED_FOR_USER : AVAILABLE;
diff --git a/src/com/android/settings/biometrics/combination/BiometricWaysToUsePreferenceController.java b/src/com/android/settings/biometrics/combination/BiometricWaysToUsePreferenceController.java
new file mode 100644
index 0000000000..99105be43f
--- /dev/null
+++ b/src/com/android/settings/biometrics/combination/BiometricWaysToUsePreferenceController.java
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.settings.biometrics.combination;
+
+import android.content.Context;
+
+import com.android.settings.Utils;
+import com.android.settings.core.BasePreferenceController;
+
+import com.android.settings.custom.biometrics.FaceUtils;
+
+public class BiometricWaysToUsePreferenceController extends BasePreferenceController {
+
+    public BiometricWaysToUsePreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return FaceUtils.isFaceUnlockSupported() ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
+    }
+}
diff --git a/src/com/android/settings/biometrics/face/FaceEnrollIntroduction.java b/src/com/android/settings/biometrics/face/FaceEnrollIntroduction.java
index 66ed56afc9..a04a0bcb59 100644
--- a/src/com/android/settings/biometrics/face/FaceEnrollIntroduction.java
+++ b/src/com/android/settings/biometrics/face/FaceEnrollIntroduction.java
@@ -22,6 +22,7 @@ import static com.android.settings.biometrics.BiometricUtils.GatekeeperCredentia
 
 import android.app.admin.DevicePolicyManager;
 import android.app.settings.SettingsEnums;
+import android.content.ComponentName;
 import android.content.Intent;
 import android.content.res.Configuration;
 import android.hardware.SensorPrivacyManager;
@@ -60,6 +61,8 @@ import com.google.android.setupcompat.template.FooterButton;
 import com.google.android.setupcompat.util.WizardManagerHelper;
 import com.google.android.setupdesign.span.LinkSpan;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Provides introductory info about face unlock and prompts the user to agree before starting face
  * enrollment.
@@ -72,6 +75,8 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
     @Nullable private FooterButton mSecondaryFooterButton;
     @Nullable private SensorPrivacyManager mSensorPrivacyManager;
 
+    private boolean mForRedo;
+
     @Override
     protected void onCancelButtonClick(View view) {
         if (!BiometricUtils.tryStartingNextBiometricEnroll(this, ENROLL_NEXT_BIOMETRIC_REQUEST,
@@ -110,28 +115,7 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
                 mDevicePostureState, mLaunchedPostureGuidance);
     }
 
-    @Override
-    protected void onCreate(Bundle savedInstanceState) {
-        mFaceManager = getFaceManager();
-
-        if (savedInstanceState == null
-                && !WizardManagerHelper.isAnySetupWizard(getIntent())
-                && !getIntent().getBooleanExtra(EXTRA_FROM_SETTINGS_SUMMARY, false)
-                && maxFacesEnrolled()) {
-            // from tips && maxEnrolled
-            Log.d(TAG, "launch face settings");
-            launchFaceSettingsActivity();
-            finish();
-        }
-
-        super.onCreate(savedInstanceState);
-
-        // Wait super::onCreated() then return because SuperNotCalledExceptio will be thrown
-        // if we don't wait for it.
-        if (isFinishing()) {
-            return;
-        }
-
+    private void initDefaultLayout(){
         // Apply extracted theme color to icons.
         final ImageView iconGlasses = findViewById(R.id.icon_glasses);
         final ImageView iconLooking = findViewById(R.id.icon_looking);
@@ -171,6 +155,34 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
             iconRequireEyes.getBackground().setColorFilter(getIconColorFilter());
             infoMessageRequireEyes.setText(getInfoMessageRequireEyes());
         }
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        mForRedo = getIntent().getBooleanExtra("for_redo", false);
+        mFaceManager = getFaceManager();
+
+        if (savedInstanceState == null
+                && !WizardManagerHelper.isAnySetupWizard(getIntent())
+                && !getIntent().getBooleanExtra(EXTRA_FROM_SETTINGS_SUMMARY, false)
+                && maxFacesEnrolled()) {
+            // from tips && maxEnrolled
+            Log.d(TAG, "launch face settings");
+            launchFaceSettingsActivity();
+            finish();
+        }
+
+        super.onCreate(savedInstanceState);
+
+        // Wait super::onCreated() then return because SuperNotCalledExceptio will be thrown
+        // if we don't wait for it.
+        if (isFinishing()) {
+            return;
+        }
+
+        if (!FaceUtils.isFaceUnlockSupported()) {
+            initDefaultLayout();
+        }
 
         // This path is an entry point for SetNewPasswordController, e.g.
         // adb shell am start -a android.app.action.SET_NEW_PASSWORD
@@ -207,6 +219,10 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
         final boolean cameraPrivacyEnabled = helper
                 .isSensorBlocked(SensorPrivacyManager.Sensors.CAMERA, mUserId);
         Log.v(TAG, "cameraPrivacyEnabled : " + cameraPrivacyEnabled);
+
+        if (FaceUtils.isFaceUnlockSupported() && mHasPassword && mToken != null) {
+            openCustomFaceUnlockPackage();
+        }
     }
 
     private void launchFaceSettingsActivity() {
@@ -304,6 +320,11 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
 
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (FaceUtils.isFaceUnlockSupported()) {
+            onActivityResultCustom(requestCode, resultCode, data);
+            return;
+        }
+
         if (requestCode == REQUEST_POSTURE_GUIDANCE) {
             mLaunchedPostureGuidance = false;
             if (resultCode == RESULT_CANCELED || resultCode == RESULT_SKIP) {
@@ -336,6 +357,66 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
         super.onActivityResult(requestCode, resultCode, data);
     }
 
+    void onActivityResultCustom(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (requestCode != CHOOSE_LOCK_GENERIC_REQUEST) {
+            if (requestCode != CONFIRM_REQUEST) {
+                if (requestCode == ENROLL_REQUEST) {
+                    if (resultCode == RESULT_FIRST_USER || resultCode == RESULT_OK) {
+                        setResult(RESULT_FIRST_USER);
+                        finish();
+                        return;
+                    }
+                    setResult(RESULT_CANCELED);
+                    finish();
+                }
+            } else if (resultCode == RESULT_OK && data != null) {
+                checkTokenAndOpenCustomFaceUnlockPackage(data);
+            }
+        } else if (resultCode == RESULT_FIRST_USER) {
+            checkTokenAndOpenCustomFaceUnlockPackage(data);
+        }
+    }
+
+    private void openCustomFaceUnlockPackage() {
+        ComponentName componentName;
+        Intent intent = new Intent();
+        intent.putExtra(ChooseLockSettingsHelper.EXTRA_KEY_CHALLENGE_TOKEN, mToken);
+        if (mUserId != -10000) {
+            intent.putExtra("android.intent.extra.USER_ID", mUserId);
+        }
+        if (mForRedo) {
+            componentName = new ComponentName(
+                "org.pixelexperience.faceunlock",
+                "org.pixelexperience.faceunlock.FaceEnrollActivity");
+        } else {
+            componentName = new ComponentName(
+                "org.pixelexperience.faceunlock",
+                "org.pixelexperience.faceunlock.SetupFaceIntroActivity");
+        }
+        intent.setComponent(componentName);
+        if (intent.resolveActivity(getPackageManager()) != null) {
+            startActivityForResult(intent, ENROLL_REQUEST);
+        }
+    }
+
+    private void checkTokenAndOpenCustomFaceUnlockPackage(Intent intent) {
+        if (mToken == null) {
+            mFaceManager.generateChallenge(mUserId, (sensorId, userId, challenge) -> {
+                if (mToken == null) {
+                    mToken = BiometricUtils.requestGatekeeperHat(this, intent, mUserId,
+                            challenge);
+                    mSensorId = sensorId;
+                    mChallenge = challenge;
+                    BiometricUtils.removeGatekeeperPasswordHandle(this, intent);
+                    openCustomFaceUnlockPackage();
+                }
+            });
+        }else{
+            openCustomFaceUnlockPackage();
+        }
+    }
+
     protected boolean generateChallengeOnCreate() {
         return true;
     }
@@ -383,6 +464,9 @@ public class FaceEnrollIntroduction extends BiometricEnrollIntroduction {
 
     @Override
     protected int getLayoutResource() {
+        if (FaceUtils.isFaceUnlockSupported()) {
+            return R.layout.face_enroll_introduction_invisible;
+        }
         return R.layout.face_enroll_introduction;
     }
 
diff --git a/src/com/android/settings/biometrics/face/FaceProfileStatusPreferenceController.java b/src/com/android/settings/biometrics/face/FaceProfileStatusPreferenceController.java
index a2e11afb7e..248dde36f6 100644
--- a/src/com/android/settings/biometrics/face/FaceProfileStatusPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceProfileStatusPreferenceController.java
@@ -27,6 +27,8 @@ import androidx.preference.Preference;
 
 import com.android.settings.R;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 public class FaceProfileStatusPreferenceController extends FaceStatusPreferenceController {
 
     private static final String KEY_FACE_SETTINGS = "face_settings_profile";
@@ -54,6 +56,9 @@ public class FaceProfileStatusPreferenceController extends FaceStatusPreferenceC
 
     @Override
     public int getAvailabilityStatus() {
+        if (FaceUtils.isFaceUnlockSupported()) {
+            return UNSUPPORTED_ON_DEVICE;
+        }
         // Check if Face for Profile is available.
         final int isAvailable = super.getAvailabilityStatus();
         if (isAvailable != AVAILABLE) {
diff --git a/src/com/android/settings/biometrics/face/FaceSettings.java b/src/com/android/settings/biometrics/face/FaceSettings.java
index dc94376b91..34b7ce5806 100644
--- a/src/com/android/settings/biometrics/face/FaceSettings.java
+++ b/src/com/android/settings/biometrics/face/FaceSettings.java
@@ -52,6 +52,9 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+import com.android.settings.custom.biometrics.face.FaceSettingsRedoPreferenceController;
+
 /**
  * Settings screen for face authentication.
  */
@@ -77,6 +80,7 @@ public class FaceSettings extends DashboardFragment {
     private FaceSettingsRemoveButtonPreferenceController mRemoveController;
     private FaceSettingsEnrollButtonPreferenceController mEnrollController;
     private FaceSettingsLockscreenBypassPreferenceController mLockscreenController;
+    private FaceSettingsRedoPreferenceController mRedoController;
     private List<AbstractPreferenceController> mControllers;
 
     private List<Preference> mTogglePreferences;
@@ -87,6 +91,12 @@ public class FaceSettings extends DashboardFragment {
     private boolean mConfirmingPassword;
 
     private final FaceSettingsRemoveButtonPreferenceController.Listener mRemovalListener = () -> {
+        if (FaceUtils.isFaceUnlockSupported()){
+            if (getActivity() != null) {
+                getActivity().finish();
+            }
+            return;
+        }
 
         // Disable the toggles until the user re-enrolls
         for (Preference preference : mTogglePreferences) {
@@ -167,9 +177,7 @@ public class FaceSettings extends DashboardFragment {
                                     R.string.security_settings_face_profile_preference_title)));
         }
 
-        mLockscreenController = Utils.isMultipleBiometricsSupported(context)
-                ? use(BiometricLockscreenBypassPreferenceController.class)
-                : use(FaceSettingsLockscreenBypassPreferenceController.class);
+        mLockscreenController = use(FaceSettingsLockscreenBypassPreferenceController.class);
         mLockscreenController.setUserId(mUserId);
 
         Preference keyguardPref = findPreference(FaceSettingsKeyguardPreferenceController.KEY);
@@ -194,6 +202,10 @@ public class FaceSettings extends DashboardFragment {
         }
         mRemoveController.setUserId(mUserId);
 
+        if (mRedoController != null) {
+            mRedoController.setUserId(mUserId);
+        }
+
         // Don't show keyguard controller for work profile settings.
         if (mUserManager.isManagedProfile(mUserId)) {
             removePreference(FaceSettingsKeyguardPreferenceController.KEY);
@@ -228,6 +240,7 @@ public class FaceSettings extends DashboardFragment {
         } else {
             mAttentionController.setToken(mToken);
             mEnrollController.setToken(mToken);
+            mRedoController.setToken(mToken);
         }
 
         final boolean hasEnrolled = mFaceManager.hasEnrolledTemplates(mUserId);
@@ -308,6 +321,9 @@ public class FaceSettings extends DashboardFragment {
                 mEnrollController = (FaceSettingsEnrollButtonPreferenceController) controller;
                 mEnrollController.setListener(mEnrollListener);
                 mEnrollController.setActivity((SettingsActivity) getActivity());
+            } else if (controller instanceof FaceSettingsRedoPreferenceController) {
+                mRedoController = (FaceSettingsRedoPreferenceController) controller;
+                mRedoController.setActivity((SettingsActivity) getActivity());
             }
         }
 
@@ -322,6 +338,7 @@ public class FaceSettings extends DashboardFragment {
         controllers.add(new FaceSettingsRemoveButtonPreferenceController(context));
         controllers.add(new FaceSettingsConfirmPreferenceController(context));
         controllers.add(new FaceSettingsEnrollButtonPreferenceController(context));
+        controllers.add(new FaceSettingsRedoPreferenceController(context));
         return controllers;
     }
 
@@ -363,6 +380,10 @@ public class FaceSettings extends DashboardFragment {
                         keys.add(FaceSettingsAttentionPreferenceController.KEY);
                     }
 
+                    if (FaceUtils.isFaceUnlockSupported()) {
+                        keys.add("security_settings_face_unlock_category");
+                    }
+
                     return keys;
                 }
 
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsAppPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsAppPreferenceController.java
index c296e56aa0..69a019c589 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsAppPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsAppPreferenceController.java
@@ -25,6 +25,7 @@ import android.provider.Settings;
 import androidx.preference.Preference;
 
 import com.android.settings.Utils;
+import com.android.settings.custom.biometrics.FaceUtils;
 
 /**
  * Preference controller for Face settings page controlling the ability to use
@@ -79,7 +80,7 @@ public class FaceSettingsAppPreferenceController extends FaceSettingsPreferenceC
     @Override
     public int getAvailabilityStatus() {
         // When the device supports multiple biometrics auth, this preference will be hidden.
-        if (Utils.isMultipleBiometricsSupported(mContext)) {
+        if (Utils.isMultipleBiometricsSupported(mContext) && !FaceUtils.isFaceUnlockSupported()) {
             return UNSUPPORTED_ON_DEVICE;
         }
 
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsAttentionPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsAttentionPreferenceController.java
index 82fa00b864..5c29e387ac 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsAttentionPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsAttentionPreferenceController.java
@@ -27,6 +27,8 @@ import androidx.preference.SwitchPreference;
 
 import com.android.settings.Utils;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller that manages the ability to use face authentication with/without
  * user attention. See {@link FaceManager#setRequireAttention(boolean, byte[])}.
@@ -118,6 +120,6 @@ public class FaceSettingsAttentionPreferenceController extends FaceSettingsPrefe
 
     @Override
     public int getAvailabilityStatus() {
-        return AVAILABLE;
+        return FaceUtils.isFaceUnlockSupported() ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
     }
 }
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsConfirmPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsConfirmPreferenceController.java
index 60f8b21d4f..df296849d5 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsConfirmPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsConfirmPreferenceController.java
@@ -30,6 +30,8 @@ import com.android.settings.Utils;
 
 import java.util.List;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller giving the user an option to always require confirmation.
  */
@@ -79,6 +81,9 @@ public class FaceSettingsConfirmPreferenceController extends FaceSettingsPrefere
 
     @Override
     public int getAvailabilityStatus() {
+        if (FaceUtils.isFaceUnlockSupported()){
+            return UNSUPPORTED_ON_DEVICE;
+        }
         List<FaceSensorProperties> properties = mFaceManager.getSensorProperties();
         // If a sensor is convenience, it is possible that it becomes weak or strong with
         // an update. For this reason, the sensor is conditionally unavailable.
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsEnrollButtonPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsEnrollButtonPreferenceController.java
index b3e0e1ea88..d7893eead0 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsEnrollButtonPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsEnrollButtonPreferenceController.java
@@ -20,8 +20,6 @@ import static com.android.settings.Utils.SETTINGS_PACKAGE_NAME;
 
 import android.content.Context;
 import android.content.Intent;
-import android.view.View;
-import android.widget.Button;
 
 import androidx.preference.Preference;
 
@@ -31,14 +29,13 @@ import com.android.settings.core.BasePreferenceController;
 import com.android.settings.password.ChooseLockSettingsHelper;
 import com.android.settingslib.widget.LayoutPreference;
 
-import com.google.android.setupdesign.util.ButtonStyler;
 import com.google.android.setupdesign.util.PartnerStyleHelper;
 
 /**
  * Preference controller that allows a user to enroll their face.
  */
 public class FaceSettingsEnrollButtonPreferenceController extends BasePreferenceController
-        implements View.OnClickListener {
+        implements Preference.OnPreferenceClickListener {
 
     private static final String TAG = "FaceSettings/Remove";
     static final String KEY = "security_settings_face_enroll_faces_container";
@@ -48,7 +45,6 @@ public class FaceSettingsEnrollButtonPreferenceController extends BasePreference
     private int mUserId;
     private byte[] mToken;
     private SettingsActivity mActivity;
-    private Button mButton;
     private boolean mIsClicked;
     private Listener mListener;
 
@@ -64,19 +60,11 @@ public class FaceSettingsEnrollButtonPreferenceController extends BasePreference
     @Override
     public void updateState(Preference preference) {
         super.updateState(preference);
-
-        mButton = ((LayoutPreference) preference).findViewById(
-                R.id.security_settings_face_settings_enroll_button);
-
-        if (PartnerStyleHelper.shouldApplyPartnerResource(mButton)) {
-            ButtonStyler.applyPartnerCustomizationPrimaryButtonStyle(mContext, mButton);
-        }
-
-        mButton.setOnClickListener(this);
+        preference.setOnPreferenceClickListener(this);
     }
 
     @Override
-    public void onClick(View v) {
+    public boolean onPreferenceClick(Preference preference) {
         mIsClicked = true;
         final Intent intent = new Intent();
         intent.setClassName(SETTINGS_PACKAGE_NAME, FaceEnrollIntroduction.class.getName());
@@ -87,6 +75,7 @@ public class FaceSettingsEnrollButtonPreferenceController extends BasePreference
         } else {
             mContext.startActivity(intent);
         }
+        return true;
     }
 
     @Override
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsKeyguardPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsKeyguardPreferenceController.java
index 342d786482..c1b79b184a 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsKeyguardPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsKeyguardPreferenceController.java
@@ -28,6 +28,8 @@ import com.android.settings.Utils;
 import com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
 import com.android.settingslib.RestrictedSwitchPreference;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller for Face settings page controlling the ability to unlock the phone
  * with face.
@@ -71,7 +73,7 @@ public class FaceSettingsKeyguardPreferenceController extends FaceSettingsPrefer
     @Override
     public int getAvailabilityStatus() {
         // When the device supports multiple biometrics auth, this preference will be unavailable.
-        return Utils.isMultipleBiometricsSupported(mContext) ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
+        return Utils.isMultipleBiometricsSupported(mContext) || FaceUtils.isFaceUnlockSupported() ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
     }
 
     @Override
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsLockscreenBypassPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsLockscreenBypassPreferenceController.java
index c0cb64ec6b..b2457b1c8b 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsLockscreenBypassPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsLockscreenBypassPreferenceController.java
@@ -83,16 +83,15 @@ public class FaceSettingsLockscreenBypassPreferenceController
 
     @Override
     public int getAvailabilityStatus() {
-        // When the device supports multiple biometrics auth, this preference won't be shown
-        // in face unlock category.
-        if (Utils.isMultipleBiometricsSupported(mContext)) {
-            return UNSUPPORTED_ON_DEVICE;
-        }
         if (mUserManager.isManagedProfile(getUserId())) {
             return UNSUPPORTED_ON_DEVICE;
         }
 
-        if (mFaceManager != null && mFaceManager.isHardwareDetected()) {
+        boolean faceAuthOnlyOnSecurityView  = mContext.getResources().getBoolean(
+                com.android.internal.R.bool.config_faceAuthOnlyOnSecurityView);
+
+        if (mFaceManager != null && mFaceManager.isHardwareDetected() &&
+                !faceAuthOnlyOnSecurityView) {
             return mFaceManager.hasEnrolledTemplates(getUserId())
                     ? AVAILABLE : DISABLED_DEPENDENT_SETTING;
         } else {
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsLockscreenUnlockMethodPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsLockscreenUnlockMethodPreferenceController.java
new file mode 100644
index 0000000000..e37f4dc8d3
--- /dev/null
+++ b/src/com/android/settings/biometrics/face/FaceSettingsLockscreenUnlockMethodPreferenceController.java
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+package com.android.settings.biometrics.face;
+
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.hardware.face.FaceManager;
+import android.os.UserHandle;
+import android.os.UserManager;
+
+import com.android.settings.core.BasePreferenceController;
+
+public class FaceSettingsLockscreenUnlockMethodPreferenceController
+        extends BasePreferenceController {
+
+    protected FaceManager mFaceManager;
+    private UserManager mUserManager;
+
+    public FaceSettingsLockscreenUnlockMethodPreferenceController(Context context, String preferenceKey) {
+        super(context, preferenceKey);
+        if (context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_FACE)) {
+            mFaceManager = context.getSystemService(FaceManager.class);
+        }
+
+        mUserManager = context.getSystemService(UserManager.class);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        if (mUserManager.isManagedProfile(UserHandle.myUserId())) {
+            return UNSUPPORTED_ON_DEVICE;
+        }
+
+        boolean faceAuthOnlyOnSecurityView  = mContext.getResources().getBoolean(
+                com.android.internal.R.bool.config_faceAuthOnlyOnSecurityView);
+
+        if (mFaceManager != null && mFaceManager.isHardwareDetected() &&
+                !faceAuthOnlyOnSecurityView) {
+            return mFaceManager.hasEnrolledTemplates(UserHandle.myUserId())
+                    ? AVAILABLE : DISABLED_DEPENDENT_SETTING;
+        } else {
+            return UNSUPPORTED_ON_DEVICE;
+        }
+    }
+}
\ No newline at end of file
diff --git a/src/com/android/settings/biometrics/face/FaceSettingsRemoveButtonPreferenceController.java b/src/com/android/settings/biometrics/face/FaceSettingsRemoveButtonPreferenceController.java
index 7db5958489..01f5196b47 100644
--- a/src/com/android/settings/biometrics/face/FaceSettingsRemoveButtonPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceSettingsRemoveButtonPreferenceController.java
@@ -25,8 +25,6 @@ import android.hardware.face.Face;
 import android.hardware.face.FaceManager;
 import android.os.Bundle;
 import android.util.Log;
-import android.view.View;
-import android.widget.Button;
 import android.widget.Toast;
 
 import androidx.preference.Preference;
@@ -40,17 +38,16 @@ import com.android.settings.overlay.FeatureFactory;
 import com.android.settingslib.core.instrumentation.MetricsFeatureProvider;
 import com.android.settingslib.widget.LayoutPreference;
 
-import com.google.android.setupdesign.util.ButtonStyler;
-import com.google.android.setupdesign.util.PartnerStyleHelper;
-
 import java.util.List;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Controller for the remove button. This assumes that there is only a single face enrolled. The UI
  * will likely change if multiple enrollments are allowed/supported.
  */
-public class FaceSettingsRemoveButtonPreferenceController extends BasePreferenceController
-        implements View.OnClickListener {
+public class FaceSettingsRemoveButtonPreferenceController extends BasePreferenceController implements
+        Preference.OnPreferenceClickListener {
 
     private static final String TAG = "FaceSettings/Remove";
     static final String KEY = "security_settings_face_delete_faces_container";
@@ -70,7 +67,7 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
             AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
 
             builder.setTitle(R.string.security_settings_face_settings_remove_dialog_title)
-                    .setMessage(mIsConvenience
+                    .setMessage(mIsConvenience && !FaceUtils.isFaceUnlockSupported()
                             ? R.string.security_settings_face_settings_remove_dialog_details_convenience
                             : R.string.security_settings_face_settings_remove_dialog_details)
                     .setPositiveButton(R.string.delete, mOnClickListener)
@@ -94,7 +91,6 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
     }
 
     private Preference mPreference;
-    private Button mButton;
     private Listener mListener;
     private SettingsActivity mActivity;
     private int mUserId;
@@ -118,7 +114,7 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
             if (remaining == 0) {
                 final List<Face> faces = mFaceManager.getEnrolledFaces(mUserId);
                 if (!faces.isEmpty()) {
-                    mButton.setEnabled(true);
+                    mPreference.setEnabled(true);
                 } else {
                     mRemoving = false;
                     mListener.onRemoved();
@@ -134,7 +130,7 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
         @Override
         public void onClick(DialogInterface dialog, int which) {
             if (which == DialogInterface.BUTTON_POSITIVE) {
-                mButton.setEnabled(false);
+                mPreference.setEnabled(false);
                 final List<Face> faces = mFaceManager.getEnrolledFaces(mUserId);
                 if (faces.isEmpty()) {
                     Log.e(TAG, "No faces");
@@ -147,7 +143,7 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
                 // Remove the first/only face
                 mFaceUpdater.remove(faces.get(0), mUserId, mRemovalCallback);
             } else {
-                mButton.setEnabled(true);
+                mPreference.setEnabled(true);
                 mRemoving = false;
             }
         }
@@ -172,21 +168,13 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
     @Override
     public void updateState(Preference preference) {
         super.updateState(preference);
-
         mPreference = preference;
-        mButton = ((LayoutPreference) preference)
-                .findViewById(R.id.security_settings_face_settings_remove_button);
-
-        if (PartnerStyleHelper.shouldApplyPartnerResource(mButton)) {
-            ButtonStyler.applyPartnerCustomizationPrimaryButtonStyle(mContext, mButton);
-        }
-
-        mButton.setOnClickListener(this);
+        mPreference.setOnPreferenceClickListener(this);
 
         if (!FaceSettings.isFaceHardwareDetected(mContext)) {
-            mButton.setEnabled(false);
+            mPreference.setEnabled(false);
         } else {
-            mButton.setEnabled(!mRemoving);
+            mPreference.setEnabled(!mRemoving);
         }
     }
 
@@ -201,15 +189,13 @@ public class FaceSettingsRemoveButtonPreferenceController extends BasePreference
     }
 
     @Override
-    public void onClick(View v) {
-        if (v == mButton) {
-            mMetricsFeatureProvider.logClickedPreference(mPreference, getMetricsCategory());
-            mRemoving = true;
-            ConfirmRemoveDialog dialog = new ConfirmRemoveDialog();
-            dialog.setOnClickListener(mOnClickListener);
-            dialog.setIsConvenience(BiometricUtils.isConvenience(mFaceManager));
-            dialog.show(mActivity.getSupportFragmentManager(), ConfirmRemoveDialog.class.getName());
-        }
+    public boolean onPreferenceClick(Preference preference) {
+        mRemoving = true;
+        ConfirmRemoveDialog dialog = new ConfirmRemoveDialog();
+        dialog.setOnClickListener(mOnClickListener);
+        dialog.setIsConvenience(BiometricUtils.isConvenience(mFaceManager));
+        dialog.show(mActivity.getSupportFragmentManager(), ConfirmRemoveDialog.class.getName());
+        return true;
     }
 
     public void setListener(Listener listener) {
diff --git a/src/com/android/settings/biometrics/face/FaceStatusPreferenceController.java b/src/com/android/settings/biometrics/face/FaceStatusPreferenceController.java
index f18a74fa19..4d809d22a3 100644
--- a/src/com/android/settings/biometrics/face/FaceStatusPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceStatusPreferenceController.java
@@ -18,6 +18,7 @@ package com.android.settings.biometrics.face;
 
 import android.content.Context;
 import android.hardware.face.FaceManager;
+import androidx.preference.Preference;
 
 import androidx.annotation.Nullable;
 import androidx.lifecycle.Lifecycle;
@@ -32,6 +33,8 @@ import com.android.settings.biometrics.BiometricStatusPreferenceController;
 import com.android.settingslib.RestrictedLockUtils;
 import com.android.settingslib.RestrictedPreference;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 public class FaceStatusPreferenceController extends BiometricStatusPreferenceController
         implements LifecycleObserver {
 
@@ -90,6 +93,9 @@ public class FaceStatusPreferenceController extends BiometricStatusPreferenceCon
     public void updateState(Preference preference) {
         super.updateState(preference);
         updateStateInternal();
+        if (FaceUtils.isFaceUnlockSupported()) {
+            preference.setEnabled(!FaceUtils.isFaceDisabledByAdmin(mContext));
+        }
     }
 
     private void updateStateInternal() {
diff --git a/src/com/android/settings/biometrics/face/FaceStatusUtils.java b/src/com/android/settings/biometrics/face/FaceStatusUtils.java
index 80ffb03b0b..5f0b1658a8 100644
--- a/src/com/android/settings/biometrics/face/FaceStatusUtils.java
+++ b/src/com/android/settings/biometrics/face/FaceStatusUtils.java
@@ -26,6 +26,8 @@ import com.android.settings.Utils;
 import com.android.settings.biometrics.ParentalControlsUtils;
 import com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Utilities for face details shared between Security Settings and Safety Center.
  */
@@ -62,6 +64,9 @@ public class FaceStatusUtils {
      * Returns the summary of face settings entity.
      */
     public String getSummary() {
+        if (FaceUtils.isFaceUnlockSupported() && FaceUtils.isFaceDisabledByAdmin(mContext)) {
+            return mContext.getResources().getString(R.string.disabled_by_administrator_summary);
+        }
         return mContext.getResources().getString(hasEnrolled()
                 ? R.string.security_settings_face_preference_summary
                 : R.string.security_settings_face_preference_summary_none);
diff --git a/src/com/android/settings/biometrics/face/FaceUnlockCategoryPreferenceController.java b/src/com/android/settings/biometrics/face/FaceUnlockCategoryPreferenceController.java
index d0debdfe59..870756ef29 100644
--- a/src/com/android/settings/biometrics/face/FaceUnlockCategoryPreferenceController.java
+++ b/src/com/android/settings/biometrics/face/FaceUnlockCategoryPreferenceController.java
@@ -20,6 +20,8 @@ import android.content.Context;
 import com.android.settings.Utils;
 import com.android.settings.core.BasePreferenceController;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * Preference controller that controls the face unlock features to be shown / be hidden.
  */
@@ -31,6 +33,6 @@ public class FaceUnlockCategoryPreferenceController extends BasePreferenceContro
 
     @Override
     public int getAvailabilityStatus() {
-        return Utils.isMultipleBiometricsSupported(mContext) ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
+        return Utils.isMultipleBiometricsSupported(mContext) && !FaceUtils.isFaceUnlockSupported() ? UNSUPPORTED_ON_DEVICE : AVAILABLE;
     }
 }
diff --git a/src/com/android/settings/custom/ListPreference.java b/src/com/android/settings/custom/ListPreference.java
new file mode 100644
index 0000000000..74c38aa260
--- /dev/null
+++ b/src/com/android/settings/custom/ListPreference.java
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2017-2018 AICP
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.custom;
+
+import android.content.Context;
+import android.text.TextUtils;
+import android.util.AttributeSet;
+
+public class ListPreference extends androidx.preference.ListPreference {
+    private boolean mAutoSummary = false;
+
+    public ListPreference(Context context, AttributeSet attrs, int defStyle) {
+        super(context, attrs, defStyle);
+    }
+
+    public ListPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public ListPreference(Context context) {
+        super(context);
+    }
+
+    @Override
+    public void setValue(String value) {
+        super.setValue(value);
+        if (mAutoSummary || TextUtils.isEmpty(getSummary())) {
+            setSummary(getEntry(), true);
+        }
+    }
+
+    @Override
+    public void setSummary(CharSequence summary) {
+        setSummary(summary, false);
+    }
+
+    private void setSummary(CharSequence summary, boolean autoSummary) {
+        mAutoSummary = autoSummary;
+        super.setSummary(summary);
+    }
+}
diff --git a/src/com/android/settings/custom/SecureSettingListPreference.java b/src/com/android/settings/custom/SecureSettingListPreference.java
new file mode 100644
index 0000000000..f9b909671e
--- /dev/null
+++ b/src/com/android/settings/custom/SecureSettingListPreference.java
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2017-2018 AICP
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.custom;
+
+import android.content.Context;
+import android.text.TextUtils;
+import android.util.AttributeSet;
+
+public class SecureSettingListPreference extends ListPreference {
+
+    public SecureSettingListPreference(Context context, AttributeSet attrs, int defStyle) {
+        super(context, attrs, defStyle);
+        setPreferenceDataStore(new SecureSettingsStore(context.getContentResolver()));
+    }
+
+    public SecureSettingListPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        setPreferenceDataStore(new SecureSettingsStore(context.getContentResolver()));
+    }
+
+    public SecureSettingListPreference(Context context) {
+        super(context);
+        setPreferenceDataStore(new SecureSettingsStore(context.getContentResolver()));
+    }
+
+    @Override
+    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
+        // This is what default ListPreference implementation is doing without respecting
+        // real default value:
+        //setValue(restoreValue ? getPersistedString(mValue) : (String) defaultValue);
+        // Instead, we better do
+        setValue(restoreValue ? getPersistedString((String) defaultValue) : (String) defaultValue);
+    }
+
+}
diff --git a/src/com/android/settings/custom/SecureSettingsStore.java b/src/com/android/settings/custom/SecureSettingsStore.java
new file mode 100644
index 0000000000..85a25ec00a
--- /dev/null
+++ b/src/com/android/settings/custom/SecureSettingsStore.java
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2017 AICP
+ * Copyright (C) 2018 CarbonROM
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.custom;
+
+import android.content.ContentResolver;
+import android.preference.PreferenceDataStore;
+import android.provider.Settings;
+
+public class SecureSettingsStore extends androidx.preference.PreferenceDataStore
+        implements PreferenceDataStore {
+
+    private ContentResolver mContentResolver;
+
+    public SecureSettingsStore(ContentResolver contentResolver) {
+        mContentResolver = contentResolver;
+    }
+
+    public boolean getBoolean(String key, boolean defValue) {
+        return getInt(key, defValue ? 1 : 0) != 0;
+    }
+
+    public float getFloat(String key, float defValue) {
+        return Settings.Secure.getFloat(mContentResolver, key, defValue);
+    }
+
+    public int getInt(String key, int defValue) {
+        return Settings.Secure.getInt(mContentResolver, key, defValue);
+    }
+
+    public long getLong(String key, long defValue) {
+        return Settings.Secure.getLong(mContentResolver, key, defValue);
+    }
+
+    public String getString(String key, String defValue) {
+        String result = Settings.Secure.getString(mContentResolver, key);
+        return result == null ? defValue : result;
+    }
+
+    public void putBoolean(String key, boolean value) {
+        putInt(key, value ? 1 : 0);
+    }
+
+    public void putFloat(String key, float value) {
+        Settings.Secure.putFloat(mContentResolver, key, value);
+    }
+
+    public void putInt(String key, int value) {
+        Settings.Secure.putInt(mContentResolver, key, value);
+    }
+
+    public void putLong(String key, long value) {
+        Settings.Secure.putLong(mContentResolver, key, value);
+    }
+
+    public void putString(String key, String value) {
+        Settings.Secure.putString(mContentResolver, key, value);
+    }
+
+}
diff --git a/src/com/android/settings/custom/biometrics/FaceUtils.java b/src/com/android/settings/custom/biometrics/FaceUtils.java
new file mode 100644
index 0000000000..b83aec83a9
--- /dev/null
+++ b/src/com/android/settings/custom/biometrics/FaceUtils.java
@@ -0,0 +1,31 @@
+package com.android.settings.custom.biometrics;
+
+import android.app.admin.DevicePolicyManager;
+import android.content.Context;
+import android.os.SystemProperties;
+import android.util.Log;
+
+import com.android.internal.util.custom.faceunlock.FaceUnlockUtils;
+
+public final class FaceUtils {
+    private static final String TAG = "FaceUtils";
+
+    public static boolean isFaceUnlockSupported() {
+        return FaceUnlockUtils.isFaceUnlockSupported();
+    }
+
+    public static boolean isFaceDisabledByAdmin(Context context) {
+        DevicePolicyManager devicePolicyManager = (DevicePolicyManager) context.getSystemService("device_policy");
+        try {
+            if (devicePolicyManager.getPasswordQuality(null) > 32768) {
+                return true;
+            }
+        } catch (SecurityException e) {
+            Log.e(TAG, "isFaceDisabledByAdmin error:", e);
+        }
+        if ((devicePolicyManager.getKeyguardDisabledFeatures(null) & 128) != 0) {
+            return true;
+        }
+        return false;
+    }
+}
diff --git a/src/com/android/settings/custom/biometrics/face/FaceEnrollActivity.java b/src/com/android/settings/custom/biometrics/face/FaceEnrollActivity.java
new file mode 100644
index 0000000000..b4ad610724
--- /dev/null
+++ b/src/com/android/settings/custom/biometrics/face/FaceEnrollActivity.java
@@ -0,0 +1,34 @@
+package com.android.settings.custom.biometrics.face;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.os.Bundle;
+import com.android.settings.biometrics.face.FaceEnrollIntroduction;
+import com.android.settings.custom.biometrics.FaceUtils;
+
+public class FaceEnrollActivity extends Activity {
+    private byte[] mToken;
+    @Override
+    public void onCreate(Bundle bundle) {
+        super.onCreate(bundle);
+        if (FaceUtils.isFaceUnlockSupported()) {
+            if (mToken == null) {
+                mToken = getIntent().getByteArrayExtra("hw_auth_token");
+            }
+            Intent faceIntroIntent = getFaceIntroIntent();
+            faceIntroIntent.putExtra("for_redo", getIntent().getBooleanExtra("for_redo", false));
+            faceIntroIntent.putExtra("hw_auth_token", mToken);
+            if (getCallingActivity() != null) {
+                faceIntroIntent.setFlags(33554432);
+            }
+            startActivity(faceIntroIntent);
+        }
+        finish();
+    }
+
+    private Intent getFaceIntroIntent() {
+        Intent intent = new Intent(this, FaceEnrollIntroduction.class);
+        intent.addFlags(268468224);
+        return intent;
+    }
+}
diff --git a/src/com/android/settings/custom/biometrics/face/FaceSettingsRedoPreferenceController.java b/src/com/android/settings/custom/biometrics/face/FaceSettingsRedoPreferenceController.java
new file mode 100644
index 0000000000..0d04770e1d
--- /dev/null
+++ b/src/com/android/settings/custom/biometrics/face/FaceSettingsRedoPreferenceController.java
@@ -0,0 +1,138 @@
+package com.android.settings.custom.biometrics.face;
+
+import android.app.AlertDialog;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.hardware.face.Face;
+import android.hardware.face.FaceManager;
+import android.util.Log;
+import android.widget.Toast;
+import androidx.preference.Preference;
+import com.android.settings.R;
+import com.android.settings.SettingsActivity;
+import com.android.settings.core.BasePreferenceController;
+import com.android.settings.slices.SliceBackgroundWorker;
+import com.android.settings.biometrics.face.FaceSettings;
+import com.android.settings.custom.biometrics.FaceUtils;
+import java.util.List;
+
+public class FaceSettingsRedoPreferenceController extends BasePreferenceController implements
+        Preference.OnPreferenceClickListener {
+    static final String KEY = "security_settings_face_redo_face_scan";
+    private static final String TAG = "FaceSettings/Redo";
+    private SettingsActivity mActivity;
+    private final Context mContext;
+    private final FaceManager mFaceManager;
+    private final FaceManager.RemovalCallback mRemovalCallback;
+    private int mUserId;
+    private byte[] mToken;
+
+    @Override
+    public String getPreferenceKey() {
+        return KEY;
+    }
+
+    public FaceSettingsRedoPreferenceController(Context context, String str) {
+        super(context, str);
+        mRemovalCallback = new FaceManager.RemovalCallback() {
+            @Override
+            public void onRemovalError(Face face, int i, CharSequence charSequence) {
+                Log.e(FaceSettingsRedoPreferenceController.TAG, "Unable to remove face: " + face.getBiometricId() + " error: " + i + " " + ((Object) charSequence));
+                Toast.makeText(mContext, charSequence, 0).show();
+            }
+
+            @Override
+            public void onRemovalSucceeded(Face face, int i) {
+                if (i == 0) {
+                    Log.v(FaceSettingsRedoPreferenceController.TAG, "onRemovalSucceeded ");
+                    Intent intent = new Intent("com.android.settings.intent.action.FACE_ENROLL");
+                    intent.putExtra("for_face", true);
+                    intent.putExtra("for_redo", true);
+                    intent.putExtra("hw_auth_token", mToken);
+                    intent.addFlags(268435456);
+                    mContext.startActivity(intent);
+                    return;
+                }
+                Log.v(FaceSettingsRedoPreferenceController.TAG, "Remaining: " + i);
+            }
+        };
+        mContext = context;
+        mFaceManager = (FaceManager) context.getSystemService(FaceManager.class);
+    }
+
+    public FaceSettingsRedoPreferenceController(Context context) {
+        this(context, KEY);
+    }
+
+    public void setUserId(int i) {
+        mUserId = i;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        super.updateState(preference);
+        if (!FaceSettings.isFaceHardwareDetected(mContext) ||
+              !mFaceManager.hasEnrolledTemplates(mUserId)) {
+            preference.setEnabled(false);
+        } else {
+            preference.setEnabled(true);
+            preference.setOnPreferenceClickListener(this);
+        }
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return FaceUtils.isFaceUnlockSupported() ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public boolean onPreferenceClick(Preference preference) {
+        showFaceRedoWarningDialog();
+        return true;
+    }
+
+    public void setActivity(SettingsActivity settingsActivity) {
+        mActivity = settingsActivity;
+    }
+
+    private void deleteFace() {
+        List enrolledFaces = mFaceManager.getEnrolledFaces(mUserId);
+        if (enrolledFaces.isEmpty()) {
+            Log.e(TAG, "No faces");
+            return;
+        }
+        if (enrolledFaces.size() > 1) {
+            Log.e(TAG, "Multiple enrollments: " + enrolledFaces.size());
+        }
+        mFaceManager.remove((Face) enrolledFaces.get(0), mUserId, mRemovalCallback);
+    }
+
+    void showFaceRedoWarningDialog() {
+        AlertDialog.Builder builder = new AlertDialog.Builder(mActivity);
+        builder.setTitle(R.string.security_settings_face_unlock_redo_face_scan_title)
+            .setMessage(R.string.face_redo_warning_msg)
+            .setPositiveButton(R.string.face_redo_confirm_btn, new DialogInterface.OnClickListener() {
+                @Override
+                public void onClick(DialogInterface dialogInterface, int i) {
+                    deleteFace();
+                }
+            })
+            .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
+                @Override
+                public void onClick(DialogInterface dialogInterface, int i) {
+                }
+            })
+            .setOnCancelListener(new DialogInterface.OnCancelListener() {
+                @Override
+                public void onCancel(DialogInterface dialogInterface) {
+                }
+            });
+        builder.create().show();
+    }
+
+    public void setToken(byte[] token) {
+        mToken = token;
+    }
+}
diff --git a/src/com/android/settings/homepage/contextualcards/FaceReEnrollDialog.java b/src/com/android/settings/homepage/contextualcards/FaceReEnrollDialog.java
index a8239241bf..f558d70540 100644
--- a/src/com/android/settings/homepage/contextualcards/FaceReEnrollDialog.java
+++ b/src/com/android/settings/homepage/contextualcards/FaceReEnrollDialog.java
@@ -33,6 +33,8 @@ import com.android.settings.Utils;
 import com.android.settings.biometrics.face.FaceUpdater;
 import com.android.settings.homepage.contextualcards.slices.FaceSetupSlice;
 
+import com.android.settings.custom.biometrics.FaceUtils;
+
 /**
  * This class is used to show a popup dialog for {@link FaceSetupSlice}.
  */
@@ -42,6 +44,7 @@ public class FaceReEnrollDialog extends AlertActivity implements
     private static final String TAG = "FaceReEnrollDialog";
 
     private static final String BIOMETRIC_ENROLL_ACTION = "android.settings.BIOMETRIC_ENROLL";
+    private static final String BIOMETRIC_ENROLL_ACTION_CUSTOM = "com.android.settings.intent.action.FACE_ENROLL";
 
     private FaceManager mFaceManager;
     private FaceUpdater mFaceUpdater;
@@ -113,7 +116,9 @@ public class FaceReEnrollDialog extends AlertActivity implements
                     return;
                 }
                 // Send user to the enroll flow.
-                final Intent reEnroll = new Intent(BIOMETRIC_ENROLL_ACTION);
+                final Intent reEnroll = new Intent(
+                    FaceUtils.isFaceUnlockSupported() ?
+                    BIOMETRIC_ENROLL_ACTION_CUSTOM : BIOMETRIC_ENROLL_ACTION);
                 final Context context = getApplicationContext();
 
                 try {
-- 
2.34.1

